Initial implementation.

dmitrizagidulin opened the pull request:
This PR updates the existing `ed25519-verification-key-2018` library to use the new 2020 spec.

~~Currently, only uses the isomorphic `@stablelib/ed25519` lib for crypto. If we need to, we can add a separate PR to add native Node.js crypto.~~
(Native crypto now added.)

dmitrizagidulin committed:
Initial implementation.

dmitrizagidulin committed:
Add jsigs compatibility matrix.

dlongley commented:
> If we need to, we can add a separate PR to add native Node.js crypto.

Yes, we'll need to do that.

aljones15 committed:
Use crypto interface across browser and node.

aljones15 committed:
Use crypto interface across browser and node.

Add utils from 2018 key to enable node 25519 keys.

Use add ported interfaces for node ed25519.

Add publicBytes to secretBytes.

Use DER encodings for verify and sign.

Add webpack tools to devDeps.

Remove node forge (#5)

* Use DER Buffer over node-forge.

* Remove forge from public key der encode.

* Remove node-forge.

* Use KEY_PREFIX at top of file.

* Use hexstring to save space in PUBLIC DER PREFIX.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use slice over getting ArrayBuffer.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use hex string from PRIVATE DER PREFIX.

* Add zeroBuffer to PUBLIC DER PREFIX.

* Use `instanceof Uint8Array` instead of `Buffer.isBuffer`.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use instanceof Uint8Array instead of isBuffer for private key material.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

Use key over crypto to prevent confusion with crypto lib.

Add no-crypto workaround for sub-modules.

Add webpack output to config.

Rename key to ed25519.

aljones15 committed:
Add utils from 2018 key to enable node 25519 keys.

aljones15 committed:
Use add ported interfaces for node ed25519.

aljones15 committed:
Add publicBytes to secretBytes.

aljones15 committed:
Use DER encodings for verify and sign.

aljones15 committed:
Add webpack tools to devDeps.

aljones15 committed:
Remove node forge (#5)

* Use DER Buffer over node-forge.

* Remove forge from public key der encode.

* Remove node-forge.

* Use KEY_PREFIX at top of file.

* Use hexstring to save space in PUBLIC DER PREFIX.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use slice over getting ArrayBuffer.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use hex string from PRIVATE DER PREFIX.

* Add zeroBuffer to PUBLIC DER PREFIX.

* Use `instanceof Uint8Array` instead of `Buffer.isBuffer`.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use instanceof Uint8Array instead of isBuffer for private key material.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Add github actions and node-forge again.

aljones15 committed:
Remove node-forge from crypto.js too.

aljones15 committed:
Move node specific DER stuff to crypto.js.

aljones15 committed:
Delete lib util.js.

aljones15 committed:
Add jsdoc strings to DerEncode functions.

aljones15 committed:
Further comment comment node crypto file.

aljones15 committed:
Add copyrights to new files.

aljones15 committed:
Remove empty line from crypto-browser.js.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove empty line from crypto.js.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Add compatibility spec.

aljones15 committed:
Remove typedArrayToBuffer & add Buffer.from to ToDer methods.

aljones15 committed:
Rename crypto to ed25519 and use key at the module name.

aljones15 committed:
Use buffer.indexOf default value of 0.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use buffer.indexOf default of 0.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove privateKeyMaterial var.

aljones15 committed:
Use ed25519 over key as module name.

aljones15 committed:
Make jsdoc description for privateKeyDerEncode less implementation specific.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Expand jsdoc comment to include format.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Include file extension in es6 import.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove Buffer.from from DerEncoding methods.

aljones15 committed:
Move randomBytes into ed25519 files.

aljones15 committed:
Remove secureRandom.

aljones15 committed:
Ensure randomBytes is a function.

aljones15 committed:
Use seedBytes over privateKeyBytes.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove privateBytes from generate.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Change parameter of verify to match parameters of publicKeyDerEncode.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Change parameters of sign to match privateKeyDerEncode.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use randomBytes in generateKeyPair method.

aljones15 committed:
Move randomBytes into respective env specific modules.

aljones15 committed:
Remove export of privateKeyBuffer in favor of using seedBytes.

aljones15 committed:
Use Uint8Array as jsdoc type over Buffer.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use byte over bit in jsdoc doc strings.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Clarify jsdoc return type for ed25519 key.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Move helpers function below api defs.

  - Create a helper function for randomBytes for @stablelib.
  - Use variable for seed in node.

aljones15 committed:
Use window.crypto.getRandomValues to generate seed to @stabelib ed25519.

aljones15 committed:
Promisify randomBytes to make async.

aljones15 committed:
Move Export line up.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Export object as default.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove helper function in favor of direct crypto call.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use import on util & move export default api up.

aljones15 committed:
Correct documentation to use Uint8Array over Buffer.

aljones15 committed:
Add tests for stable verifying node and vice versa.

aljones15 committed:
Add test for node key verification on both back and front end.

aljones15 committed:
Add tests that run in browser and node using base58 encoded signatures.

aljones15 committed:
Remove copy and paste comments from mock-data.

aljones15 committed:
Remove @stabelib/random.

aljones15 committed:
Upgrade karma to 5 to avoid security vulnerability in 4.

davidlehn commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,73 @@
+    - run: npm install
+    - name: Run eslint
+      run: npm run lint
+#  coverage:
```
`nyc` is being installed, should probably add the coverage commands to use it and enable coverage here.

dmitrizagidulin replied:
@davidlehn let's open a separate issue/PR for that.

dlongley reviewed (approved):
Approving for my part here, over to @davidlehn for any remaining webpack/build issues.

dmitrizagidulin commented:
Merging since approved; will fine-tune webpack issues in separate PR. (Preliminary downstream testing with vc-js suggests that the webpack works.)
